### ISSUE TITLE
BackoffSupervisor max number of retries not working as expected

### ISSUE THREAD

**Issue by tometchy** [akkadotnet/akka.net#4601]
Hi everyone!

Firstly for the record:
- Which Akka.Net: **1.4.8**
- On which platform you are using Akka.Net: **NetCore 2.1 (C# 7.3)**
- Gist which can be easily used to reproduce the case: **https://gist.github.com/tometchy/40c712c7d5be7071f1f55d697da9af22**

I have taken such assumptions: _BackoffSupervisor_ will try to recreate child actor until max backoff time will elapse, but not more number of attempts than specified in _maxNrOfRetries_ argument.
So for example with _maxBackoff_ being _TimeSpan.MaxValue_ and _maxNrOfRetries_ being _1_ it will recreate only once.

If above assumptions are wrong please correct me. If they are correct, please look at provided gist in which I show that it doesn't work this way.

From logs output you can read that child actor was created 7 times (_[Actor for one request created]_), despite max number of retries was 5, so it should be created only 6 times (first initial time and 5 retries).

**Comment by ismaelhamed** [comment]
Your assumption is correct, and we have tests for that: https://github.com/akkadotnet/akka.net/blob/59e149df45538040d424d29ef1586fbe63ffd603/src/core/Akka.Tests/Pattern/BackoffSupervisorSpec.cs#L348

In this case, the problem is that you are overriding the default supervision strategy without setting the `maxNrOfRetries`, which will default to `-1` (notice that `maxNrOfRetries` is actually a property of the `SupervisionStrategy`, not the `BackOffSupervisor` itself).

Use this in your code and it should work as expected:

`.WithSupervisorStrategy(new OneForOneStrategy(MaxNumberOfRetries, null, exception => Directive.Restart));`

Better yet, with a `BackOff.OnStop` use the `.WithDefaultStoppingStrategy` to always stop the actor, since the BackOffSupervisor already uses its own "restart" procedure.

**Comment by tometchy** [comment]
I see, there is space for ux improvement here :)

But unfortunately this doesn't solve the issue, both with `.WithSupervisorStrategy(new OneForOneStrategy(MaxNumberOfRetries, null, exception => Directive.Restart)));` and `.WithDefaultStoppingStrategy());`.

Actually, with `.WithDefaultStoppingStrategy());` it's even worse, it doesn't work as you described - it doesn't restart actor at all.

How I have changed code from gist:
``` csharp
BackoffSupervisor.Props(
                Backoff.OnStop(Props.Create(() => new RequestActor()), 
                        childName: nameof(RequestActor),
                        minBackoff: TimeSpan.FromSeconds(2),
                        maxBackoff: MaxBackOff,
                        randomFactor: 0.2,
                        maxNrOfRetries: MaxNumberOfRetries)
                    // .WithSupervisorStrategy(new OneForOneStrategy(MaxNumberOfRetries, null, exception => Directive.Restart)));
                    .WithDefaultStoppingStrategy());
```

Logs outputs:
``` text
2020-11-02_10:51:26.052 [Debug] [Logger log1-NLogLogger [NLogLogger] started] | 
2020-11-02_10:51:26.087 [Debug] [StandardOutLogger being removed] | 
2020-11-02_10:51:26.087 [Debug] [Default Loggers started] | 
2020-11-02_10:51:26.087 [Info] [Became idle] | 
2020-11-02_10:51:26.087 [Debug] [Got download command, creating request actor] | 
2020-11-02_10:51:26.087 [Debug] [Telling request actor to download] | 
2020-11-02_10:51:26.087 [Info] [Became working] | 
2020-11-02_10:51:26.087 [Debug] [Actor for one request created] | 
2020-11-02_10:51:26.087 [Info] [Got download command] | 
2020-11-02_10:51:26.087 [Warn] [Request failed] | 
2020-11-02_10:51:26.087 [Debug] [Telling request actor to download] | 
2020-11-02_10:51:26.087 [Error] [Couldn't download] | EXCEPTION OCCURRED:System.Exception Couldn't download Boolean <.ctor>b__0(Download)    at Phone.Service.Tests.ExternalQueues.ProveBug.RequestActor.<>c__DisplayClass0_0.<.ctor>b__0(Download update) in C:\GIT\EQ\src\Server\Phone.Service.Tests\ExternalQueues\ExternalQueuesAggregateTests.cs:line 152
   at lambda_method(Closure , Object , Func`2 )
   at Akka.Actor.ReceiveActor.ExecutePartialMessageHandler(Object message, PartialAction`1 partialAction)
   at Akka.Actor.UntypedActor.Receive(Object message)
   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   at Akka.Actor.ActorCell.ReceiveMessage(Object message)
   at Akka.Actor.ActorCell.Invoke(Envelope envelope)
2020-11-02_10:51:26.126 [Info] [Message [Download] from akka://test/user/$a to akka://test/user/$a/RequestActor/RequestActor was not delivered. [1] dead letters encountered .This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.] | 
2020-11-02_10:51:28.796 [Debug] [Received download command while already working, skipping: "Phone.Service.Tests.ExternalQueues.ProveBug.DownloaderActor+Do+Download"] |
```

You see, actor for request got created only once (_[Actor for one request created]_) and when it told to sender about failure and sender sent _Download_ message again, it ended up in dead letters (_Message [Download] from akka://test/user/$a to akka://test/user/$a/RequestActor/RequestActor was not delivered. [1] dead letters encountered_).

I have even tried to remove this _.WithXStrategy_ entirely, but it didn't help neither.

**Comment by ismaelhamed** [comment]
There must be something off with that code, I'll try to take a closer look. 

Btw, `maxBackoff` means that the exponential back-off is capped to this duration (interval between restarts is bounded by `minBackoff` and `maxBackoff`), not that restarts won't happen after `maxBackoff`.

**Comment by tometchy** [comment]
Oh, thanks for clarification. So for example if min backoff is 1s and maxbackoff is 4s and number of retries is 5, then we will have retries with such delays, am I right (random factor skipped in this example for simplification):
wait 1s then retry
wait 2s then retry
wait 4s then retry
wait 4s then retry
wait 4s then retry
End of retries (number of retries 5 reached).

**Comment by ismaelhamed** [comment]
Well, since you have also specified a `randomFactor`, the exponential back-off will take that into acount too:

wait 1s * 0.2 then retry,
wait 2s * 0.2 then retry, 
...

but basically, yes.

**Comment by tometchy** [comment]
I took investigation again and found the issue. You were right, `.WithSupervisorStrategy(new OneForOneStrategy(MaxNumberOfRetries, null, exception => Directive.Restart));` is the solution.

I got mislead because I had another (wrong) assumption - I thought that such supervisor kills himself after reaching max number of retries.
After investigation I realized (finally), that he didn't.
He has to be killed manually, and created manually again when needed, because it can't no longer be used, as it has reached it's recreating limit. So he is in state which makes him useless, but is still alive.

Once again space for UX improvement here :)

Still thank you very much for your guidance, finally I have resolved this dread simple case.
